tests/test_0184-concatenate-operation.py

```diff
@@ -0,0 +1,179 @@
+    two = awkward1.Array([7, 99, 999, 8, 9]).mask[[True, False, False, True, True]].layout
+
+    assert awkward1.to_list(awkward1.concatenate([one, two], 0)) == [1, 2, None, 4, None, 6, 7, None, None, 8, 9]
+
+    with pytest.raises(ValueError) as err:
+        awkward1.to_list(awkward1.concatenate([one, two], 1))
+
+def test_listoffsetarray_concatenate():
+    content_one = awkward1.layout.NumpyArray(numpy.array([1, 2, 3, 4, 5, 6, 7, 8, 9]))
+    offsets_one = awkward1.layout.Index64(numpy.array([0, 3, 3, 5, 9]))
+    one = awkward1.layout.ListOffsetArray64(offsets_one, content_one)
+
+    assert awkward1.to_list(one) == [[1, 2, 3], [], [4, 5], [6, 7, 8, 9]]
+
+    content_two = awkward1.layout.NumpyArray(numpy.array([100, 200, 300, 400, 500]))
+    offsets_two = awkward1.layout.Index64(numpy.array([0, 2, 4, 4, 5]))
+    two = awkward1.layout.ListOffsetArray64(offsets_two, content_two)
+
+    assert awkward1.to_list(two) == [[100, 200], [300, 400], [], [500]]
+    assert awkward1.to_list(awkward1.concatenate([one, two], 0)) == [[1, 2, 3], [], [4, 5], [6, 7, 8, 9], [100, 200], [300, 400], [], [500]]
+    assert awkward1.to_list(awkward1.concatenate([one, two], 1)) == [[1, 2, 3, 100, 200], [300, 400], [4, 5], [6, 7, 8, 9, 500]]
+
+def test_numpyarray_concatenate_axis0():
+    emptyarray = awkward1.layout.EmptyArray()
+
+    np1 = numpy.arange(2*7*5, dtype=numpy.float64).reshape(2, 7, 5)
+    np2 = numpy.arange(3*7*5, dtype=numpy.int64).reshape(3, 7, 5)
+    ak1 = awkward1.layout.NumpyArray(np1)
+    ak2 = awkward1.layout.NumpyArray(np2)
+
+    assert awkward1.to_list(awkward1.concatenate([np1, np2, np1, np2], 0)) == awkward1.to_list(numpy.concatenate([np1, np2, np1, np2], 0))
+    assert awkward1.to_list(awkward1.concatenate([ak1, ak2], 0)) == awkward1.to_list(numpy.concatenate([ak1, ak2], 0))
+    assert awkward1.to_list(numpy.concatenate([ak1, ak2], 0)) == awkward1.to_list(awkward1.concatenate([np1, np2], 0))
+
+def test_numpyarray_concatenate():
+
+    np1 = numpy.arange(2*7*5, dtype=numpy.float64).reshape(2, 7, 5)
+    np2 = numpy.arange(2*7*5, dtype=numpy.int64).reshape(2, 7, 5)
+    ak1 = awkward1.layout.NumpyArray(np1)
+    ak2 = awkward1.layout.NumpyArray(np2)
+
+    assert awkward1.to_list(numpy.concatenate([np1, np2], 1)) == awkward1.to_list(awkward1.concatenate([ak1, ak2], 1))
+    assert awkward1.to_list(numpy.concatenate([np2, np1], 1)) == awkward1.to_list(awkward1.concatenate([ak2, ak1], 1))
+    assert awkward1.to_list(numpy.concatenate([np1, np2], 2)) == awkward1.to_list(awkward1.concatenate([ak1, ak2], 2))
+    assert awkward1.to_list(numpy.concatenate([np2, np1], 2)) == awkward1.to_list(awkward1.concatenate([ak2, ak1], 2))
+
+def test_numbers_and_records_concatenate():
+
+    numbers = [
+        awkward1.Array([[1.1, 2.2, 3.3], [], [4.4, 5.5], []]),
+        awkward1.Array([[10, 20], [30], [], [40, 50, 60]]),
+        awkward1.Array(numpy.array([[101, 102, 103], [201, 202, 203], [], []])),
```

Comment:
So, `broadcast_and_apply` _should_ be requiring all input arrays to have equal lengths at all levels _before_ you apply your customization (through `getfunction`). So if `axis=10`, all of the input arrays would have to have the same lengths for 10 levels of recursion, and then any lengths they have deeper than that do not matter. For instance, with `axis=1`, all input arrays must have the same length (4 in this example). If `axis=2`, then they must all have the same outer array length as well as all nested list lengths, but then they can have whatever they want deeper than that.

These arrays can only do `axis=0` and `axis=1`. Looking closely, yes, I goofed up all their lengths. The first one has length 3, the second has length 4, and so you get the error message. Here's a better set:

```python
        numbers = [
            awkward1.Array([
                [1.1, 2.2, 3.3],
                [],
                [4.4, 5.5],
                [6.6, 7.7, 8.8, 9.9],
            ]),
            awkward1.Array([
                [10, 20],
                [30],
                [],
                [40, 50, 60],
            ]),
            awkward1.Array(numpy.array([
                [101, 102, 103],
                [201, 202, 203],
                [301, 302, 303],
                [401, 402, 403],
            ])),
        ]
        records = [
            awkward1.Array([
                [{"x": 0.0, "y": []}, {"x": 1.1, "y": [1]}],
                [],
                [{"x": 2.2, "y": [1, 2]}],
                [{"x": 3.3, "y": [1, 2, 3]}, {"x": 4.4, "y": [1, 2, 3, 4]}],
            ]),
            awkward1.Array([
                [{"x": 0.0, "y": []}],
                [{"x": 1.1, "y": [1]}, {"x": 2.2, "y": [1, 2]}],
                [],
                [{"x": 3.3, "y": [1, 2, 3]}],
            ]),
        ]
```

They all have length 4. The lengths within each list differ wildly, but `broadcast_and_apply` should only do its length check for the first level when `axis=1`.

Here's what you should get from `ak.concatenate([numbers[0], numbers[1]], axis=1)`:

```python
[
    [1.1, 2.2, 3.3, 10, 20],
    [30],
    [4.4, 5.5],
    [6.6, 7.7, 8.8, 9.9, 40, 50, 60],
]
```

And here's what you should get from `ak.concatenate([numbers[0], records[0]], axis=1)`:

```python
[
    [1.1, 2.2, 3.3, {"x": 0.0, "y": []}, {"x": 1.1, "y": [1]}],
    [],
    [4.4, 5.5, {"x": 2.2, "y": [1, 2]}],
    [6.6, 7.7, 8.8, 9.9, {"x": 3.3, "y": [1, 2, 3]}, {"x": 4.4, "y": [1, 2, 3, 4]}],
]
```

If you want an example that tests `broadcast_and_apply`'s length checking at more than one level, there's this:

```python
arrays = [
    ak.Array([[[0.0, 1.1, 2.2], []], [[3.3, 4.4]], [[5.5], [6.6, 7.7, 8.8, 9.9]]]),
    ak.Array([[[10, 20], [30]], [[40]], [[50, 60, 70], [80, 90]]]),
]
# nothing is required to have the same length
assert ak.concatenate(arrays, axis=0).tolist() == [
    [[0.0, 1.1, 2.2], []],
    [[3.3, 4.4]],
    [[5.5], [6.6, 7.7, 8.8, 9.9]],
    [[10, 20], [30]],
    [[40]],
    [[50, 60, 70], [80, 90]],
]
# the outermost arrays are required to have the same length, but nothing deeper than that
assert ak.concatenate(arrays, axis=1).tolist() == [
    [[0.0, 1.1, 2.2], [], [10, 20], [30]],
    [[3.3, 4.4], [40]],
    [[5.5], [6.6, 7.7, 8.8, 9.9], [50, 60, 70], [80, 90]],
]
# the outermost arrays and the first level are required to have the same length, but nothing deeper
assert ak.concatenate(arrays, axis=2).tolist() == [
    [[0.0, 1.1, 2.2, 10, 20], [30]],
    [[3.3, 4.4, 40]],
    [[5.5, 50, 60, 70], [6.6, 7.7, 8.8, 9.9, 80, 90]],
]
```

You should be able to concatenate `[arrays[0][:, :-1], arrays[1]]` at `axis=1`, but not at `axis=2`.

Reply:
ok, the tests pass, but there is a warning:
```python
tests/test_0184-concatenate-operation.py::test_list_offset_array_concatenate PASSED                                                                                                                                                                  [ 10%]
tests/test_0184-concatenate-operation.py::test_list_array_concatenate PASSED                                                                                                                                                                         [ 20%]
tests/test_0184-concatenate-operation.py::test_records_concatenate PASSED                                                                                                                                                                            [ 30%]
tests/test_0184-concatenate-operation.py::test_indexed_array_concatenate PASSED                                                                                                                                                                      [ 40%]
tests/test_0184-concatenate-operation.py::test_bytemasked_concatenate PASSED                                                                                                                                                                         [ 50%]
tests/test_0184-concatenate-operation.py::test_listoffsetarray_concatenate PASSED                                                                                                                                                                    [ 60%]
tests/test_0184-concatenate-operation.py::test_numpyarray_concatenate_axis0 PASSED                                                                                                                                                                   [ 70%]
tests/test_0184-concatenate-operation.py::test_numpyarray_concatenate PASSED                                                                                                                                                                         [ 80%]
tests/test_0184-concatenate-operation.py::test_numbers_and_records_concatenate PASSED                                                                                                                                                                [ 90%]
tests/test_0184-concatenate-operation.py::test_jims_test PASSED                                                                                                                                                                                      [100%]

===================================================================================================================== warnings summary =====================================================================================================================
tests/test_0184-concatenate-operation.py::test_numbers_and_records_concatenate
  /Users/yana/Projects/PR184-rebase/awkward-1.0/awkward1/_util.py:64: DeprecationWarning: In version 1.0.0 (target date: 2020-12-01), this will be an error.
  (Set ak.deprecations_as_errors = True to get a stack trace now.)
  
  ValueError: cannot broadcast: <class 'awkward1._ext.RecordArray'>, <class 'awkward1._ext.RecordArray'>
  
  (https://github.com/scikit-hep/awkward-1.0/blob/0.4.4/src/awkward1/_util.py#L884)
    warnings.warn(message, DeprecationWarning)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
============================================================================================================== 10 passed, 1 warning in 0.77s ===============================================================================================================

```
I've also removed all kernels I wrote for this PR to make sure this code was not causing the dereference error. The PR is pretty minimal now. It relies on the `offsets_and_flattened` and `broadcast_and_apply`. I can try to move this PR tests after the 0397 test, but apart from that I'm running out of the ideas on how to catch the dereference bug. Could it be that the garbage collector is simply delayed?

Reply:
fixed